File: code/0002.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode()
        ori_head = head
        carry = 0
        while True:
            l1_val = l1.val if l1 is not None else 0
            l2_val = l2.val if l2 is not None else 0
            sum = l1_val + l2_val + carry
            head.val = sum % 10
            carry = sum // 10
            head.next = ListNode()
            l1 = l1.next if l1 is not None else l1
            l2 = l2.next if l2 is not None else l2
            if(l1 is None and l2 is None):
                if carry == 0:
                    head.next = None
                else:
                    head.next.val = carry
                break
            head = head.next
        return ori_head

File: code/0002.cpp
// Codes for 
//     2. Add Two Numbers
// Author : Yu-Cheng Lin 
//          2021. 12. 20 - version 1 
// Complexity : O(n)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        short quo = 0, rem;
        bool start = true;
        ListNode *head, *cur, *test;
        while(1){
            short sum = (l1==nullptr ? 0 : l1->val) + (l2==nullptr ? 0 : l2->val) + quo;
            quo = sum / 10;
            rem = sum - quo * 10;
            bool last = (l1 == nullptr && l2 == nullptr);
            if(!start && !(last && rem==0)) 
                {cur->next = new ListNode(rem); cur->next->val = rem;}
            if(start) {cur = new ListNode(rem); head = cur; start = false;}
            else cur = cur->next;
            if(last) break;
            if(l1) l1 = l1->next;
            if(l2) l2 = l2->next;
        }
        return head;
    }
};